### ISSUE TITLE
Membership Cost

### ISSUE THREAD

**Issue by staheri14** [vacp2p/research#56]
**Insertion**: Inserting a member to the smart contract costs 40k gas = 0.00656 Eth = 14.34 USD (considering gas price as 164 Gwei, and each Eth is worth of 1715.46 USD). 

**Deletion**: The same cost of 14.34 USD applies to the member deletion. 

**Membership fee**: In addition to the cost associated with the member insertion, each peer  must lock some fund in the contract for the sake of spam protection.  The fund will be rewarded to the slasher for catching a spammer. Slashing a member involves deletion operation which costs  14.34 USD. Thus, for the slashing to make sense, the fund locked by each user must be more than the deletion cost i.e 14.34 USD. This yields the registration fee that is more than 2*14.34 = 28.68 USD.

This amount may/may not be considered reasonable for becoming a member of a spam protected messaging system.

**Comment by oskarth** [comment]
Thanks for these numbers, useful! 40k gas seems pretty low, similar to ERC20 transfer. But I agree that they are probably an order of magnitude or two off from what would be desirable. That should be doable with some L2 solution like Optimism though, or a similar route.

FWIW, the gas cost for deploying a swap contract (done once per user) is `2100308`:

```
eth_sendTransaction                                                                   
  Contract call:       SimpleSwapFactory#deploySimpleSwap                             
  Transaction:         0xf7628153fa45d6994be14035250436e0118c8df6534f00cb4f6d4cefa43f2
72b                                                                                   
  From:                0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266                     
  To:                  0xe7f1725e7734ce288f8367e1bb143e90bb3f0512                     
  Value:               0 ETH                                                          
  Gas used:            2085828 of 2100308                                             
  Block #4:            0x8b5997038f0239685e9ba886f2194468702aee14bf4ff0637a87b3ed40e68
5f0                     
```

Cashing a cheque is around 100k:

```
eth_sendTransaction                                                                   
  Contract call:       ERC20SimpleSwap#cashChequeBeneficiary                          
  Transaction:         0xe3b174088d1b9299c9e684b84b15e44609ebc92606bf856d4411a942b9074
969                                                                                   
  From:                0x70997970c51812dc3a010c7d01b50e0d17dc79c8                     
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c                     
  Value:               0 ETH                                                          
  Gas used:            102306 of 102306                                               
  Block #8:            0xb76a3b8092c9d94b9975a9b5e50365085f66aaae17ca2d87d7f7f8a488e6d
d23   
```

These can probably be optimized, but just to give a ballpark.

**Comment by oskarth** [comment]
I took the liberty of rewriting this issue a bit

**Comment by oskarth** [comment]
Here's what running it on Optimistic public testnet would look like (coming in March, supposedly):
- https://community.optimism.io/docs/evm-comparison.html#missing-replaced-and-custom-opcodes
- and https://community.optimism.io/docs/integration.html#introduction

We'd have to look over contracts to check opcodes etc.

As for other solutions, I don't have a clear view of timelines, capabilities or maturity of them.

**Comment by cyanlemons** [comment]
> As for other solutions, I don't have a clear view of timelines, capabilities or maturity of them.

@oskarth Aztec's Noir language for private smart contracts [is said](https://medium.com/aztec-protocol/aztec-2-0-pre-launch-notes-aabad022808d) to be available "over the coming months." Also, it seems like they are adding ECDSA support, so Keycard may not end up being a blocker.

This will likely be a more finicky & less mature solution in the short term, but in the long term, I think we may regret building infrastructure on L2s that don't prioritize privacy. And intuitively, I would presume that having a completely private financial entry point to Waku would eliminate entire classes of potential attack vectors.

**Comment by oskarth** [comment]
This is a good point, and private settlement is indeed a desideratum. For now, considering the complexity and many unknowns both around the solutions, and L2s in general, I'd probably err on the side of doing simplest L2 possible that would scale. Once that seems to be working, private settlement is indeed something that would be bumped up in priority.

It is also a question of specific threat models, where both RLN and SWAP don't directly tie two people talking together (nor need chat and crypto keys be tied together), say, but is more about paying a provider for some service. Ideally, this is private as well of course.

To sum up:
- it should be mentioned in specs in terms of security guarantees,
- starting with simplest possible thing and iterating from there,
- eyes on the long term goal of having private settlement

**Comment by oskarth** [comment]
I realize I might've misunderstood what prompted the the original issue a bit. This is really two issues that we might've been conflating:

**1) More cost effective way to do Ethereum transactions in general - measured in ETH or USD terms.**

This is where L2 and similar solutions come in. This is true for *any* kind of interaction, as even ETH and ERC20 token transfers can be high on Ethereum current L1 when there's high demand.

2) Possibly high *gas cost* for certain membership operations.

If I understand correctly, we are also implicitly saying this is also the case, independent of 1 above. When we talk about shortcomings and why made certain changes, such as moving merkle tree to local as opposed to keeping in in smart contract, the rationale we are using is that tree member insert/delete is expensive.

When looking at above numbers, I see 40k gas for inserting a member and a log cost of members. I don't know what the constant is here, but assuming 2^20 merkle tree (1M elements that's a multipler of 6 for <some multiplier>). That doesn't at its face seem unreasonable.

For comparison:
- Transfers: ETH 21k, ERC20 ~50k
- ZK operations: ECADD and ECMUL are both <1000 gas whereas pairing checks are 70-110k+ gas
- Tornado Cash is something like 800k-1m (which is definitely quite high...)

Since this has implications for construct and how you access Merkle root, I think we want to get more certainty on (2) here. Specifically, what is the most expensive tree operation (presumably insert) in gas cost for a tree of e.g. size 2^20? If it is <100k gas, I'd say this means:

- It isn't a real problem, and
- We should instead address (1) as an orthogonal problem
- We might want to consider reintroducing the Merkle tree in contract as original design

For when you are back from vacation @staheri14
